### PR TITLE
Implement favorite fragment and add/remove TextView to favorite list in the home page

### PR DESCRIPTION
**Pull request by sahar-as**
New Parts: 
- Change home Fragment Ui and add text view in order to add / remove from favorite list
- Crate favorite fragment and view model
- Use same Adapter for home and favorite page
- In Favorite fragment if select remove text view, the item will remove

### CODE REVIEW

File: app/src/main/java/ir/saharapps/rickandmorty/ui/screen/characters/CharactersAdapter.kt [hunk 1]
```diff
@@ -26,9 +30,25 @@ class CharactersAdapter(private val onClick: (id: Int) -> Unit): ListAdapter<Cha
             binding.imgCharacterPic.load(character.image)
             binding.txtCharacterName.text = character.name
 
-            binding.root.setOnClickListener {
+            val selectedCharacter = getItem(adapterPosition)
+
+            if(selectedCharacter.isFavorite){
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/ui/screen/characters/CharactersAdapter.kt:35]: It's wrong way to update ui elements, R&D and find best solution

File: app/src/main/java/ir/saharapps/rickandmorty/ui/screen/favorite/FavoriteFragment.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        binding.rvFavoriteCharacters.apply {
+            layoutManager = LinearLayoutManager(this.context)
+            adapter = favoriteAdapter
+            val dividerItemDecoration = DividerItemDecoration(this.context, LinearLayoutManager.VERTICAL)
+            addItemDecoration(dividerItemDecoration)
+        }
+
+        lifecycleScope.launch {
+            favoriteViewModel.viewStateFlow.collect{characters ->
+                favoriteAdapter.submitList(characters)
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/ui/screen/favorite/FavoriteFragment.kt:49]: Handle Empty state in favorite list

**Review by FarshidRoohi — CHANGES_REQUESTED** [review]

File: app/src/main/java/ir/saharapps/rickandmorty/domain/utility/ToCharacterExt.kt [hunk 3]
```diff
@@ -9,5 +11,8 @@ fun CharacterEntity.toCharacterUi(): Character {
         name = name,
         image = image,
         isFavorite = isFavorite,
+        favTextColor  = if(isFavorite) Color.WHITE else Color.BLACK,
+        favBackground = if (isFavorite)  R.drawable.round_dark_purple_background else R.drawable.round_light_purple_background,
+        favText = if (isFavorite) "Remove From favorite" else "Add to favorite"
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/domain/utility/ToCharacterExt.kt:16]: You can't use hard code string, it's a problem

**Review by FarshidRoohi — CHANGES_REQUESTED** [review]

File: app/src/main/java/ir/saharapps/rickandmorty/ui/screen/favorite/FavoriteFragment.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        binding.rvFavoriteCharacters.apply {
+            layoutManager = LinearLayoutManager(this.context)
+            adapter = favoriteAdapter
+            val dividerItemDecoration = DividerItemDecoration(this.context, LinearLayoutManager.VERTICAL)
+            addItemDecoration(dividerItemDecoration)
+        }
+
+        lifecycleScope.launch {
+            favoriteViewModel.viewStateFlow.collect{state ->
+                when(state.emptyFavList){
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/ui/screen/favorite/FavoriteFragment.kt:56]: It's good and work, but you can write better

File: app/src/main/java/ir/saharapps/rickandmorty/ui/screen/favorite/FavoriteFragment.kt [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                        noFaveUi(false)
+                        favoriteAdapter.submitList(state.characters)
+                    }
+                }
+            }
+        }
+    }
+
+    private fun noFaveUi(isListEmpty: Boolean){
+        if (isListEmpty){
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/ui/screen/favorite/FavoriteFragment.kt:68]: It's good and work, but you can write better

File: app/src/main/java/ir/saharapps/rickandmorty/ui/screen/favorite/FavoriteViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    private val useCase: CharactersUseCase
+): ViewModel() {
+
+    private val _viewStateFlow = MutableStateFlow(FavoriteViewState())
+    val viewStateFlow: StateFlow<FavoriteViewState> = _viewStateFlow.asStateFlow()
+
+    fun getAllFavorite(){
+        viewModelScope.launch(Dispatchers.IO) {
+            val characters = useCase.getAllFavorite()
+            if (characters.isNotEmpty()){
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/ui/screen/favorite/FavoriteViewModel.kt:28]: Improve this conditions

File: app/src/main/java/ir/saharapps/rickandmorty/ui/screen/favorite/FavoriteViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+                _viewStateFlow.updateState { copy(characters = characters) }
+            }else{
+            _viewStateFlow.updateState { copy(emptyFavList = true) }
+            }
+        }
+    }
+
+    fun updateFavoriteState(id: Int){
+        viewModelScope.launch(Dispatchers.IO) {
+            val characters = useCase.updateFavStateReturnFavCharacter(id, false)
```
Comment (FarshidRoohi) [inline app/src/main/java/ir/saharapps/rickandmorty/ui/screen/favorite/FavoriteViewModel.kt:38]: Improve your conditions

File: app/src/main/res/layout/fragment_favorite.xml [hunk 8]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    tools:context=".ui.screen.favorite.FavoriteFragment">
+
+    <TextView
+        android:id="@+id/txt_no_favorite"
```
Comment (FarshidRoohi) [inline app/src/main/res/layout/fragment_favorite.xml:10]: rename `txt_no_favorite` to `txt_empty_state`

**Review by FarshidRoohi — CHANGES_REQUESTED** [review]

**Comment by sahar-as** [comment]
Edit:
- edit changeVisibility of favFragment
- Edit text view to txtEmptyState,
- Improve condition in favViewmodel"

**Pull request merged**